openshift-kube-apiserver/admission/autoscaling/managednode/admission.go

```diff
@@ -0,0 +1,159 @@
+)
+
+const (
+	PluginName = "autoscaling.openshift.io/ManagedNode"
+	// timeToWaitForCacheSync contains the time how long to wait for caches to be synchronize
+	timeToWaitForCacheSync = 10 * time.Second
+	// infraClusterName contains the name of the cluster infrastructure resource
+	infraClusterName = "cluster"
+)
+
+var _ = initializer.WantsExternalKubeClientSet(&managedNodeValidate{})
+var _ = admission.ValidationInterface(&managedNodeValidate{})
+var _ = WantsInfraInformer(&managedNodeValidate{})
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(config io.Reader) (admission.Interface, error) {
+			return &managedNodeValidate{
+				Handler: admission.NewHandler(admission.Create, admission.Update),
+			}, nil
+		})
+}
+
+type managedNodeValidate struct {
+	*admission.Handler
+	client                kubernetes.Interface
+	infraConfigLister     configv1listers.InfrastructureLister
+	infraConfigListSynced func() bool
+}
+
+// SetExternalKubeClientSet implements the WantsExternalKubeClientSet interface.
+func (a *managedNodeValidate) SetExternalKubeClientSet(client kubernetes.Interface) {
+	a.client = client
+}
+
+func (a *managedNodeValidate) SetInfraInformer(informer configv1informer.InfrastructureInformer) {
+	a.infraConfigLister = informer.Lister()
+	a.infraConfigListSynced = informer.Informer().HasSynced
+}
+
+func (a *managedNodeValidate) ValidateInitialization() error {
+	if a.client == nil {
+		return fmt.Errorf("%s plugin needs a kubernetes client", PluginName)
+	}
+	if a.infraConfigLister == nil {
+		return fmt.Errorf("%s did not get a config infrastructure lister", PluginName)
+	}
+	if a.infraConfigListSynced == nil {
+		return fmt.Errorf("%s plugin needs a config infrastructure lister synced", PluginName)
+	}
+	return nil
+}
+
+func (a *managedNodeValidate) waitForSyncedStore(timeout <-chan time.Time) bool {
+	for !a.infraConfigListSynced() {
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case <-timeout:
+			return a.infraConfigListSynced()
+		}
+	}
+	return true
+}
+
+func (a *managedNodeValidate) Validate(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	if attr.GetResource().GroupResource() != corev1.Resource("nodes") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	node, ok := attr.GetObject().(*coreapi.Node)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	if !a.waitForSyncedStore(time.After(timeToWaitForCacheSync)) {
+		return admission.NewForbidden(attr, fmt.Errorf("%s infra config cache not synchronized", PluginName))
+	}
+
+	clusterInfra, err := a.infraConfigLister.Get(infraClusterName)
+	if err != nil {
+		return admission.NewForbidden(attr, err) // can happen due to informer latency
+	}
+
+	// the infrastructure status is empty, so we can not decide the cluster type
+	if reflect.DeepEqual(clusterInfra.Status, configv1.InfrastructureStatus{}) {
```

Comment:
does this have any adverse effect during bootstrapping or upgrade?

Reply:
In testing it does not, after giving this another look, it's not really needed at this stage and might only cause issues or confusion, we do an explicit check for the value we care about after this check so this becomes redundant. Removing it, thanks!